docs/rfcs/0003-rfc-controller_refactor.md

```diff
@@ -0,0 +1,112 @@
+
+
+## Proposed Solution
+Move all connectivity related code out of the controller and into its own separate component which is referred to as the Connectivity Manager in this document. The controller object will continue to remain but its purpose will be drastically simplified to a few basic functionalities including `activate()` and `deactivate()`. The connectivity manager will primarily act as a liaison between the controller and the UI. 
+
+```mermaid
+graph TD
+    Controller --> ConnectivityManager
+    ConnectivityManager --> Frontend
+    Frontend --> Controller
+```
+
+The existing Connection Health object in VPN performs checks that act in the same space as the connectivity manager. For example in connection health we periodically send pings and collect some data upon receiving (or not receiving) a response which allows us to draw conclusions regarding network connectivity and server availability. Because there is so much overlap in the nature of both classes, the creation and implementation of the Connectivity Manager component allows us to entirely get rid of connection health and move any remaining logic to connectivity manager. We should be able entirely eliminate the `ConnectionHealth` class and its corresponding states (Stable, Unstable, NoSignal). The remainder of logic which sends and receives pings and measures latency and standard deviation can be all moved to the `ConnectivityManager`.
+
+Beyond the connectivity checks, the Controller also manages how much time has elapsed since the VPN was toggled on. Having this logic in controller bloats the code without any gains, so I would like to move that code out of `Controller` and instead introduce a simple `Timer` object that encapsulates the related logic.
+
+Because this is all part of a unified connectivity experience, we do not want to release bits and pieces of this until everything is implemented, test and ready to go, naturally this means that all implementation will live behind a feature flag until we are ready to test and release.
+
+
+## Implementation Plan
+Introduce a new `ConnectivityManager` component that is activated when the client is launched and connects to `networkWatcher` in case anything changes. This new component will manage the following work:
+- Internet probe
+- Firewall
+- Captive portal check
+- Server probe
+- Subscription status
+
+We will also add a new `Timer` component to keep track of how much time has elapsed since the VPN was activated. 
+
+
+The `ConnectivityManager` has various connection states to keep track of which connection related task the VPN is performing at any given point. This means that the `ConnectivityManager` will communicate with the `Controller` so that the controller knows which state to enter at every stage. Here is the proposed set of `ConnectivityManager` states and how they map to the `Controller` states:
+
```

Comment:
Is it possible to make the controller *only* take connection updates from this manager? Or even completely drop the controller state? I have some concern that even if we currently write this so that only the manager updates the controller state, at some point in the future someone will write code directly updating the controller state - causing the two to get out of sync and some weird bugs.

Reply:
Yes, good call out. I think we can entirely get rid of Controller states so we don't need to keep 2 different state machines in sync